draft-ietf-tls-esni.md

```diff
@@ -388,23 +440,34 @@ ClientHelloOuter, and a digest of the complete ClientHelloInner.
 
 When sending ClientHello, the client first computes ClientHelloInner, including
 any PSK binders, and then MAY substitute extensions which it knows will be
-duplicated in ClientHelloOuter. To do so, the client computes a hash H of the
+duplicated in ClientHelloOuter. To do so, the client computes a hash of the
 entire ClientHelloInner message with the same hash as for the KDF used to
-encrypt ClientHelloInner. Then, the client removes and and replaces extensions
-from ClientHelloInner with a single "outer_extensions" extension. The list of
-outer_extensions include those which were removed from ClientHelloInner, in the
-order in which they were removed. The hash contains the full ClientHelloInner
-hash H computed above.
-
-This process is reversed by client-facing servers upon receipt. Specifically,
-the server replaces the "outer_extensions" with extensions contained in
-ClientHelloOuter. The server then computes a hash H' of the reconstructed
-ClientHelloInner. If H' does not equal OuterExtensions.hash, the server aborts
-the connection with an "illegal_parameter" alert.
+encrypt ClientHelloInner. [[TODO: Will there always be a hash function
+associated with the KDF? I.e., does HPKE use HKDF?]] Then, the client removes
+and replaces extensions from ClientHelloInner with a single "outer_extensions"
+extension. The list of `outer_extensions` include those which were removed from
+ClientHelloInner, in the order in which they were removed. The hash contains the
+full ClientHelloInner hash computed above.
+
+This process is reversed by client-facing server. Specifically,
+the server replaces the `outer_extensions` with extensions contained in
+ClientHelloOuter. The server then computes a hash of the reconstructed
+ClientHelloInner. If the hash does not equal OuterExtensions.hash, the server
+aborts the connection with an "illegal_parameter" alert.
+
+[[TODO: This procedure destroys the intended order of the extensions. Say the
+extensions in the outer CH are [a, b, c, d] and I want to shallow-copy b in the
+inner CH. What I would send is I send [a, b, c, encrypted_client_hello = [a, c,
+d, outer_extensions = [ shallow_copy_of_b ]]]. The client-facing server doesn't
+know the order in which the client intended them to be delivered.]]
+
+[[TODO: What does the hash accomplish? It seems superfluous to me.]]
```

Comment:
It makes sure that the outer extension contents are not modified.

Reply:
Is there a specific attack this prevents?

Reply:
I don't think so. We didn't have analyze this case (yet), so the safe thing was just hash to prevent changes. On reflection, any change in the outer extension values would change the transcript if the inner CH was selected, so maybe it's not needed? This is worth an issue.

Reply:
Done. https://github.com/tlswg/draft-ietf-tls-esni/issues/263